### DISCUSSION TITLE
Upgrading to Scala 2.13 and compiler warnings

### DISCUSSION THREAD

**Discussion by oschrenk** [slick/slick#2428] [General]
I'm running into the issue of compiler warnings when generating Slick 3.3.3 code with Scala 2.13.8

1)
```
[error] /acme/src/main/scala/de/acme/persistence/schema/Tables.scala:4:23: early initializers are deprecated; they will be replaced by trait parameters in 3.0, see the migration guide on avoiding var/val in traits.
[error] object Tables extends {
```

That was addressed in this PR 
See also https://github.com/slick/slick/pull/2081

But I also see

2)
```                   ^
[error] /acme/src/main/scala/de/acme/persistence/schema/Tables.scala:464:41: Unused import
[error]   import slick.collection.heterogeneous._
[error]                                         ^
[error] /acme/src/main/scala/de/acme/persistence/schema/Tables.scala:465:48: Unused import
[error]   import slick.collection.heterogeneous.syntax._
[error]                                                ^

```

Is see 1) being addressed in  v3.4.0-M1 . See also discussion here https://github.com/slick/slick/discussions/2344

But since I also rely on [Slickless](https://github.com/underscoreio/slickless) I am kind of stuck in terms of upgrading to Scala 2.13 (without a bigger undertaking)

Do you have any recommendations? I asked slickless if they intend to [support 3.4](https://github.com/underscoreio/slickless/issues/122) but in the meantime, I would also be happy to emit a `@nowarn`  for a custom slick code compiler.  Is there a way to manipulate the outer "shell" of the generated code?

PS
I would also be happy with other ways of suppressing compiler warning for that single file but I am not aware of a way to do that beyond `@nowarn`.

**Comment by nafg** [comment]
If you want you could give https://github.com/nafg/slick-additions#slick-additions-codegen a try. It's a different code generator, using scalameta ASTs instead of strings.

But if you just want to silence warnings for specific files you can do that with compiler flags. See https://www.scala-lang.org/2021/01/12/configuring-and-suppressing-warnings.html. Specifically, you can use the `src` filter to filter by path, and the `site` filter to filter by fully-qualified name. They both seem to use regex syntax. For example, I found this added to `scalacOptions` in a project of mine: `"-Wconf:cat=lint-byname-implicit&site=eyewear\\.models.*:is"`

**Comment by oschrenk** [comment]
Thank you. I truly learned something. I remember looking for ways to configure the compiler warning and for some reason (weak search terms on my end I would presume) I never found the docs to configure warnings properly. That should already help a lot (on this topic and other issues with migrating an older Scala 2.12 base to Scala 2.13). 

As for the alternative compiler - another good idea that is worth exploring (after the Scala 2.13 migration)

**Comment by oschrenk** [comment]
This is what I use now

```
scalacOptions ++= Seq(
  "-Wconf:cat=lint-multiarg-infix&site=acme\.persistence\\.schema\\..*:s",
  "-Wconf:cat=deprecation&src=persistence/schema/.*:s",
  "-Wconf:cat=unused&site=acme\.persistence\\.schema\\..*:s",
  )

```

Figuring these flags out and especially being confused by `deprecation` and it not working with `site` qualifier took me quite a few compilation cycles. this endeavour might have helped me rationalize getting a new laptop. So much heat, fan noise and just waiting.
